OSOE-125: Add JavaScript processing functionality

0liver opened the pull request:
OSOE-125

Commit:
Add .gitattributes with * text=auto

Commit:
Reorder scripts, remove SCSS_TARGET from lint:styles scripts

Commit:
Adding ESLint to the pack

Commit:
Generate warning when eslint finds a problem

Commit:
Move ESLint config into property, use --source parameter with fallback value

Commit:
Add clean:scripts and transpile:script (doesn't work yet)

Commit:
Add logging prefixes for better readability

Commit:
Trying to make babel.config.json overridable

Commit:
Not so fast: keep config file, but revert override

 -  babel searches for the file provided by the consumer via --babel=babel.config.json in the node_modules directory, not the consumer one

Commit:
Make eslint config actually work

Commit:
Rename transpile to compile for simplicity

Commit:
Adding minification to the mix. Hurray!

Commit:
Fix file includes

Commit:
Add lint all script

Commit:
More logging

Commit:
Fix inconsistent path traversal behavior on win and linux

Commit:
Add documentation for JavaScript pipeline

 - adjust ESLint configuration paths

Commit:
Fix path

Commit:
Adding checks for paths before calling JS scripts

Piedone commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+* text=auto
```
Use the latest of this: https://github.com/Lombiq/Open-source-template/blob/dev/.gitattributes

Piedone commented on Lombiq.NodeJs.Extensions/Docs/JavaScript.md:
```diff
@@ -0,0 +1,84 @@
+
+The usage of these scripts is optional. `Lombiq Node.js Extensions` is able to process your SCSS files during the regular build of your project via MSBuild or the `dotnet` CLI without further work from your side. These scripts are meant to be used during development for short feedback loops, most of all the `watch` script.
+
+Looking for something similar for .NET? Check out our [.NET Analyzers project](https://github.com/Lombiq/.NET-Analyzers).
```
```suggestion
The scripts below lint (with [ESLint](https://eslint.org/)), transpile to ES5, and minify the given JS files into an output folder. Beyond that, there are also `clean` and `watch` scripts.

The usage of these scripts is optional. `Lombiq Node.js Extensions` is able to process your JS files during the regular build of your project via MSBuild or the `dotnet` CLI without further work from your side. These scripts are meant to be used during development for short feedback loops, most of all the `watch` script.

Looking for something similar for .NET? Check out our [.NET Analyzers project](https://github.com/Lombiq/.NET-Analyzers).
```

Piedone commented on Lombiq.NodeJs.Extensions/Docs/JavaScript.md:
```diff
@@ -0,0 +1,84 @@
+- *eslint.lombiq-base.js*: These rules are Lombiq overrides for [airbnb-base](https://www.npmjs.com/package/eslint-config-airbnb-base).
+- *eslint.config.js*: In this file you can define your own overriding rules.
+
+### Customizing ESLint rules - coming soon!
```
Please add this too. This should cover what's under https://github.com/Lombiq/Gulp-Extensions/blob/dev/Docs/JsTargets.md#eslint-rules.

Piedone commented on Lombiq.NodeJs.Extensions/Docs/JavaScript.md:
```diff
@@ -0,0 +1,84 @@
+
+The default paths for JS input and output files are *Assets/Scripts* and *wwwroot/js*, respectively. The existing folder structure in the input folder will be mirrored in the output, e.g. *Assets/Scripts/app/main.js* will be transformed into *wwwroot/js/app/main.js*, together with *wwwroot/js/app/main.min.js* and *wwwroot/js/app/main.js.map*.
+
+### Overriding the defaults - coming soon!
```
Please add this too.

Piedone commented on Lombiq.NodeJs.Extensions/Docs/JavaScript.md:
```diff
@@ -0,0 +1,84 @@
+
+### Integration with Visual Studio (Code)
+
+Visual Studio support ESLint out of the box. You can enable it by ticking the checkbox "Enable ESLint" under *Tools -> Options -> Text Editor -> JavaScript/TypeScript -> Linting -> General*.
```
This doesn't fully replicate https://github.com/Lombiq/Gulp-Extensions/blob/dev/Docs/JsTargets.md#:~:text=If%20you%20use,the%20lines%20of%3A

0liver replied:
The remainder of the linked section is obsolete now.

Piedone replied:
Previously, VS needed a separate file in a specific location for JS linting to work. Isn't that the case today?

0liver replied:
I've got it to work at one point, but after a full clean, it's broken again. So, yes, it's possible without extra files, I just have to make it permanent.

Piedone replied:
OK, let's see it then.

0liver replied:
The solution-wide `.eslintrc` might be a nice feature, but it has doubled my amount of grey hair during the past days. I've now come to the conclusion that non-project level configuration needs the upfront build of GE, which I **really** wanted to ditch with NE. I'm also now wondering if we couldn't get away with all NE-dependent projects simply extending the lombiq base rules individually. What do you think?

Piedone replied:
We need some kind of common ESLint config since we don't want to have to maintain config overrides in N places. Due to the reusable and shared nature of all OSOCE projects it's not necessarily exactly the same here, but generally in all solutions we need a central, single config for ESLint. This might be achievable in different ways though, not just what we've done before.

Piedone replied:
Why does an upfront build needs to happen though? Since NE is referenced by the projects it's used from, it should be at the root of the dependency tree, being built first.

0liver replied:
I've realized in the meantime, that the problematic part of the whole configuration setup is **only** with the VS ESLint integration, not with running ESLint from NE during the build itself :grin

So, yes, no upfront build should be unnecessary.

0liver replied:
I've got the VS ESLint support working now for JS files in the Samples project, with both locations for the _.eslintrc_ file, in a project, and inside the solution root folder.

For some reason it doesn't work yet in the NuGet samples, yet. I'm working on it.

Piedone commented on Lombiq.NodeJs.Extensions/pnpm-lock.yaml:
```diff
@@ -1,8 +1,17 @@
 lockfileVersion: 5.3
 
 specifiers:
+  '@babel/cli': ^7.17.10
```
After a build, the pnpm-lock files of all three projects (this one, UI Kit, Privacy) show up as changed in Git. There's no actually visible change, and when stating them they vanish:

https://user-images.githubusercontent.com/1976647/171283072-9570747a-75d4-4120-b8c1-45f66ea74122.mp4

This is the same as with SCSS files built into wwwroot folders kept in the repos. I suspect it's a line ending issue despite it not showing up as a diff. Perhaps nomalizing line Git line endings will help with this, or removing and reading the files.

Note that this only happens after the first build of a cleaned repo.

0liver replied:
I've seen this too, and have been unable to pinpoint the root cause 🤔

0liver replied:
Looks like [`core.safecrlf`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-coresafecrlf) set to `false` would achieve what we need here.

But I am currently unable to test this, because the annoying change is somehow gone 😮

Piedone replied:
Did you also close VS, clean the repo, and build after that?

0liver replied:
Yes, all of it.

Piedone commented on Lombiq.NodeJs.Extensions/Docs/JavaScript.md:
```diff
@@ -0,0 +1,84 @@
+
+For historical reasons, Windows uses the `\r\n` character combination (also known as CR-LF) to denote a line break, while Unix-like operating systems such as Linux and macOS simply use a single `\n` character (LF). Git (made by the creator of Linux) treats the Unix-style line endings as the only right option. If you are on Windows your Git client is almost certainly configured to "Checkout Window ...
+
+To ensure that the files have consistent line endings in the remote repository, you can add the following _.gitattributes_ file:
```
```suggestion
To see the different configurations of default and non-default paths in action, please check out our dedicated [Samples](../../Lombiq.NodeJs.Extensions.Samples/Readme.md) [projects](../../Lombiq.NodeJs.Extensions.Samples.NuGet/Readme.md).

### Integration with Visual Studio (Code)

Visual Studio supports ESLint out of the box. You can enable it by ticking the checkbox "Enable ESLint" under *Tools → Options → Text Editor → JavaScript/TypeScript → Linting → General*.

To use ESLint in Visual Studio Code, you can use e.g. Microsoft's official [ESLint plugin](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint).


## ESLint rules

The rules are found in 2 files:
- *eslint.lombiq-base.js*: These rules are Lombiq overrides for [airbnb-base](https://www.npmjs.com/package/eslint-config-airbnb-base).
- *eslint.config.js*: In this file you can define your own overriding rules.

### Customizing ESLint rules - coming soon!


## Operating System Compatibility Regarding Git and Line Breaks

For historical reasons, Windows uses the `\r\n` character combination (also known as CR-LF) to denote a line break, while Unix-like operating systems such as Linux and macOS simply use a single `\n` character (LF). Git (made by the creator of Linux) treats the Unix-style line endings as the only right option. If you are on Windows your Git client is almost certainly configured to "Checkout Windows-style, commit Unix-style" by default to overcome this cultural difference, but if not then it's a good practice to [configure Git](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration#_formatting_and_whitespace) to ensure your line endings are consistent. We've disabled the [`linebreak-style`](https://eslint.org/docs/rules/linebreak-style) rule to avoid cross compatibility issues.

To ensure that the files have consistent line endings in the remote repository, you can add the following _.gitattributes_ file:
```

0liver replied:
Sorry, I didn't commit your suggestion here. I had changed the text in the meantime, so the suggestion was out of date before I saw it.

Piedone replied:
No problem, but please change the still applicable parts accordingly.

Piedone commented on Lombiq.NodeJs.Extensions/Docs/JavaScript.md:
```diff
@@ -0,0 +1,84 @@
+# Scripts for JS files
```
Extend the normal and NuGet sample with JS processing.

0liver replied:
Should I do anything with the compiled, minified JS files, like use them on a page? Or can I spare that part?

Piedone replied:
No need for that.

Piedone commented on Lombiq.NodeJs.Extensions/package.json:
```diff
@@ -3,55 +3,78 @@
-    "__shell_docs__": "https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-shell/tests/shell.test.ts"
+    "---": "------------------------------------------------- SCRIPTS ------------------------------------------------",
+    "build:scripts": "run-p --print-label --continue-on-error lint:scripts minify:scripts",
+    "lint:scripts":
+      "(TARGET_DIR=$(node scripts/get-target) && node scripts/exists $TARGET_DIR/Assets/Scripts || exit 0 && cd $TARGET_DIR npx eslint Assets/Scripts --max-warnings 0 --config ${INIT_CWD}/config/eslint.config.json) || AREA=scripts STEP=linting pnpm _error",
+    "minify:scripts": "run-s --print-label compile:scripts _minify:scripts",
+    "compile:scripts":
+      "(TARGET_DIR=$(node scripts/get-target) && node scripts/exists $TARGET_DIR/Assets/Scripts || exit 0 && cd $TARGET_DIR && babel Assets/Scripts --out-dir wwwroot/js --config-file ${INIT_CWD}/config/babel.config.json) || AREA=scripts STEP=compiling pnpm _error ",
```
These lines are really long, break them up a bit.

0liver replied:
I've shortened them, but there's no way to break them up. It'd be possible to defer parts of the call chain to another script entry, but every invocation comes with a not-so-small cost, so I'm trying to avoid unnecessary calls.

Piedone replied:
Why is there no way to break them?

0liver replied:
Because in JSON, keys and values must be string constants.

Piedone replied:
I see, of course.

0liver replied:
In this longest line of all (showing the current version) I'm still considering two things:
```json
"compile:scripts":
  "SRC=$(node scripts/get-path js source) DEST=$(node scripts/get-path js target) && babel $SRC --out-dir $DEST --config-file ${INIT_CWD}/config/babel.config.json || AREA=scripts STEP=compilation pnpm error",
```

1. Renaming the `scripts` folder to `_` which would be shorter, somehow OK due to `underscore.js` using the same, and those scripts being internal, anyways. That would save **12 chars** here.
```json
"compile:scripts":
   "SRC=$(node _/get-path js source) DEST=$(node _/get-path js target) && babel $SRC --out-dir $DEST --config-file ${INIT_CWD}/config/babel.config.json || AREA=scripts STEP=compilation pnpm error",
```
2. Moving _babel.config.json_ into the parent folder, which would be `Lombiq.NodeJs.Extensions`' root folder. Then, I wouldn't have to pass the path to the config explicitly.
```json
"compile:scripts":
   "SRC=$(node scripts/get-path js source) DEST=$(node scripts/get-path js target) && babel $SRC --out-dir $DEST || AREA=scripts STEP=compilation pnpm error",
``` 
3. Moving _babel.config.json_  into the consumer project, just like _.stylelintrc_ and _.eslintrc_. This would have the same effect as 2. with the addition that it would allow to easily opt out of transpilation to ES2015 by removing parts of its content. I've wanted to do this initially, but then deemed it out of scope.

What do you think?

Piedone replied:
I wouldn't do 1 since it hurts understandability. 3 sounds good instead of 2.

0liver replied:
Great! I agree 😀

0liver replied:
I'm refraining from changing this right now because the babel config file also contains settings for the sitemap generation and I wouldn't like that to be exposed to the consumer. So I'm leaving this part as is.

Piedone commented on Lombiq.NodeJs.Extensions/scripts/minify.js:
```diff
@@ -0,0 +1,43 @@
+
+        const destination = path.basename(file, '.js') + '.min.js';
+
+        // Prefer the terser CLI for simplity over the API. See https://github.com/terser/terser for more information.
```
```suggestion
        // Prefer the terser CLI for simplicity over the API. See https://github.com/terser/terser for more information.
```

Commit:
Update .gitattributes to the new default

0liver committed:
Improve docs in JavaScript.md

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

Commit:
Rename and format ESLint configuration files

Commit:
Copy .eslintrc into consumer project for VS support

Commit:
Update docs

Commit:
Fix ESLint execution incl. correct .eslintrc lookup

Commit:
Formatting

Commit:
Update docs

Commit:
Update packages to latest, add if-env

Commit:
Allow lines up to 150 chars

Commit:
Make JS source dir overridable

Commit:
Keep NE npm package up to date during development by removing old files before copying the current ones

Commit:
temp changes

Commit:
Generalize path retrieval, refactor JS scripts

Commit:
Clean up, restore functionality

Commit:
Add comments to sample

Commit:
Make the JS pipeline work with custom paths in Samples project

Commit:
Generate MSBuild error in case compilation or minification fails

 - refactor remaining error code

Commit:
Consolidate .gitignore

Commit:
Extract deletion into separate Target, call before Clean

Commit:
Add watch functionality

Commit:
Prevent linter violations from happening to fix build

Commit:
Turn off line-break style rule

0liver committed:
Fix typo

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

Commit:
Apply arrow suggestion

Commit:
Rename sample script file

Commit:
Also delete files in the npm package

Commit:
Set up JS sample in NuGet.Samples

Commit:
Duplicate packages into peerDependencies to enable ESLint to find the necessary configs and plugins

Commit:
Comment console statement to fix build error

Commit:
Comment console statement to fix build error in NuGet Samples, too

Commit:
Remove unnecessary peerDependencies

Commit:
Add instructions on how to get VS ESLint support working (again)

Commit:
Sample errors

Commit:
Add description, remove .eslintrc.json from Samples

Commit:
Update docs with path overriding

Commit:
Update NE to latest v0.0.75

Commit:
Extend scripts in Samples

Commit:
Improve .eslintrc template, set root=true to get ESLint in VS to work in NuGet samples

Commit:
Fix some bugs, comment violations

Commit:
Add support for nested folders to minify script

Commit:
Cleaning up. Scripts are not ready to be combined yet.

Commit:
Fix watch:scripts, use relative paths for shorter commands

Commit:
Add generated JS files to DLL resources, too

Commit:
Make compile script the outer one, make optional

Commit:
Fix sourceMap URL

Commit:
Remove unused items

Commit:
Trying to embed resources correctly

Commit:
Update NuGet package description

Commit:
Trying even harder to get EmbeddedResources right

Commit:
Adding dependency on new target OrchardCoreEmbedModuleAssets

 - this should fix the problem with the target build order (OrchardCoreEmbedModuleAssets before AddGeneratedFilesToEmbeddedResourceList)

Commit:
Revert "Trying even harder to get EmbeddedResources right"

This reverts commit 8c02b2ce90f1e1021a1df476064f395f90ab832d.

Commit:
Update version number

Commit:
Update NE to latest

Commit:
Update npm packages

Commit:
Trying shameful hoisting to create shorter paths

Commit:
Sync docs with JS

Commit:
Revert "Trying shameful hoisting to create shorter paths"

This reverts commit ac48b0455b9607319b599bb76cff1e30a1b872e9.

The problem is that hoisting does NOT change the directory hierarchy - it simply adds more symlinks at higher levels.

Commit:
Remove unneeded variable, add explicit pnpm version

Commit:
Trying to get eslint to find the airbnb config

Commit:
Clarify NuGetBuild dependency

Commit:
Fix ESLint errors due to missing config or plugin

Commit:
Update after switch to pnpm@7.3.0

Commit:
Copy global .eslintrc to desired directory

Commit:
Fix IsValid condition

Commit:
Fix condition

Commit:
Use pnpm@7 to fix path problems

Commit:
Get all files included under Linux, too

Commit:
Fix warnings in files

Commit:
Remove noise from the build output

Commit:
Fix "stylelint-config-standard-scss" not being found

Commit:
Update description

Commit:
Copy .eslintrc from the NE installation directory

Commit:
Rename .eslintrc.json to be added to NuGet package

Commit:
Re-add missing files to NuGet package

Commit:
Update documentation

Commit:
Downgrade to latest pnpm@6 due to changes in pnpm@7

- which break our script invocations!

Commit:
Remove unneeded dependencies

Commit:
Trying to fix "The input line is too long. The syntax of the command is incorrect." - again

Commit:
Remove invalid option

Commit:
Fix scripts execution in NuGet sample project

Commit:
Remove unused pnpm-lock.yaml

Commit:
Remove unneeded packageManager property

Commit:
Make project level eslint dependencies work!

Commit:
Remove obsolete configuration :-)

Commit:
Make global eslint dependencies work!

Commit:
Fix error when property empty, shorten Command, format

Commit:
Make NE work from NuGet package

Commit:
Need to have the package.json, after all

Commit:
Add package.json to solution root, too, if desired

Commit:
Remove obsolete suppressions

Commit:
Only run the NodeJsExtensionsGlobalESLintConfigurationDirectory if it's a valid path

Commit:
Finally an NE version that does it all!

Commit:
Unify paths - trying to fix broken UIKit build

Commit:
Update documentation

Commit:
Only copy package.json and call pnpm install if a project level config is desired

Commit:
Improve documentation

Commit:
Add more logging for failing call

Commit:
Add more logging in the right place

Commit:
Use latest v6 of pnpm to fix build problem

Commit:
Even more logging, please!

Commit:
Update NuGet package, ignore failing corepack prepare command for TC

Commit:
Lower log levels back again

Commit:
Update NE NuGet package

Piedone commented on Lombiq.NodeJs.Extensions/Docs/JavaScript.md:
````diff
@@ -40,42 +40,64 @@ To use the `npm` scripts defined in this project, add any or all of the followin
+The *.eslintrc.json* file initially extends *.eslintrc.lombiq-base.js* from the Node.js Extensions `npm` package. It will automatically be created in your project during the first build. Should you prefer to use a global *.eslintrc.json* file for your whole solution, you can instruct Node.js Extensions to create that file in the location specified by the MSBuild property `<NodeJsExtensionsGlobalE ...
+
+```xml
+<NodeJsExtensionsCreateESLintConfigurationFile>$(MSBuildThisFileDirectory)</NodeJsExtensionsCreateESLintConfigurationFile>
````
This and `NodeJsExtensionsCreateStylelintConfigurationFile` take a different approach to the same problem. Let's settle on one approach and use this in both cases (i.e. looking at Directory.Build.props, one should see something like `SomeStylelintConfigForNodJsExtensions` and `SomeESLintConfigForNodJsExtensions` working exactly in the same fashion).

0liver replied:
I agree, and this will happen as part of finishing OSOE-84. I've introduced these inconsistencies as improvements to the previous solution that I build for the SCSS pipeline. On the other hand, I felt that refactoring the latter was out of scope for the issue at hand.

Piedone replied:
Does https://lombiq.atlassian.net/browse/OSOE-84 include a note about this?

0liver replied:
Now it does.

Piedone commented on Lombiq.NodeJs.Extensions/config/consumer/package.project.json:
```diff
@@ -0,0 +1,16 @@
+  "scripts": {
+    "build": "npm explore nodejs-extensions -- pnpm build",
+    "clean": "npm explore nodejs-extensions -- pnpm clean",
+    "watch": "npm explore nodejs-extensions -- pnpm watch"
```
I get these errors when starting `watch` and it indeed doesn't rebuild on change (but repeats this crash):

![image](https://user-images.githubusercontent.com/1976647/175826828-6b20a5c2-7e13-41ef-a684-58b6498bd589.png)

Manual build works.

0liver replied:
This is the breaking change I mentioned [here](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/pull/124#discussion_r918845962). It should work fine with **pnpm@v6**.

Piedone commented on Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.csproj:
```diff
@@ -54,14 +53,64 @@
+      Text="The path '$(NodeJsExtensionsGlobalESLintConfigurationDirectory)' in 'NodeJsExtensionsGlobalESLintConfigurationDirectory' is not a directory."
+      Condition="'$(NodeJsExtensionsGlobalESLintConfigurationDirectory)' != '' AND '$(IsGlobalESLintConfigurationDirectoryValid)' != 'true'" />
+    <!-- We need pnpm, so we enable it here explicitly using `corepack enable` from Node.js >=16.9. -->
+    <Exec Command="corepack enable" IgnoreExitCode="true"
+          Condition="'$(IsGlobalESLintConfigurationDirectoryValid)' == 'true'" />
+    <!-- We use the latest pnpm@6 version, which sadly cannot be expressed more succinctly.-->
+    <Exec Command="corepack prepare pnpm@6.32.24 --activate" IgnoreExitCode="true"
+          Condition="'$(IsGlobalESLintConfigurationDirectoryValid)' == 'true'" />
```
It seems to me that these take care of the PNPM dependency and thus manual installation as per the Readme is not necessary anymore.

0liver replied:
The consumer of NE could still be using Node.js with a version < 16.9, where `corepack` does not exist yet. In that case, these calls will fail, which will be ignored, but as long as pnpm is installed, our script should still work.

Piedone replied:
We can require that Node version, if it's available on GitHub Actions, there's no issue installing it in TC, and team members can do that too (https://lombiq.atlassian.net/browse/ADHOC-158).

0liver replied:
Just because we can, doesn't mean we should.

As it currently stands, NE will be usable even for lower Node.js versions as long as `pnpm` will be available. Why would we want to break that?

Piedone replied:
Fair enough. The Readme asks you to enable corepack though, which is not necessary.

0liver replied:
I've removed that part and added the version restriction for `pnpm`.

Piedone commented on Lombiq.NodeJs.Extensions/config/consumer/.eslintrc.global.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Shouldn't this file and .eslintrc.project.json kept in sync? If yes, add a comment about it in both. Same for package.json files.

0liver replied:
I don't feel there's much to keep in sync for the `.eslintr.*.json` files.

I've added a comment to the `package.*.json` files, though.

Piedone replied:
There's not yet. But if we happen to change any of them, we'll need to remember to consider the other one too.

Piedone commented on Lombiq.NodeJs.Extensions/scripts/minify.js:
```diff
@@ -1,43 +1,60 @@
+
+const workingDir = process.cwd();
+
+console.debug(`Minifying files in "${workingDir}" ...`);
+
+const fsItems = [];
+// Walk the directory tree.
+walk(workingDir)
+    .on('data', (item) => fsItems.push(item))
+    .on('end', () => {
+        fsItems.forEach((item) => {
+            const isFile = item.stats.isFile();
+            const filePath = item.path;
+
+            // We only care for .js files.
+            if (!isFile || !filePath.endsWith('.js') || filePath.endsWith('.min.js')) return;
+
+            const destination = path.join(path.dirname(filePath), path.basename(filePath, '.js') + '.min.js');
+            const sourceMapUrl = path.basename(destination) + '.map';
+
+            // Prefer the terser CLI for simplicity over the API. See https://github.com/terser/terser for details.
+            const command = [
+                'terser',
+                `"${filePath}"`,
+                `--output "${destination}"`,
+                '--compress',
+                '--mangle',
+                `--source-map "content=inline,url='${sourceMapUrl}'"`,
+            ].join(' ');
+
+            exec(command, (err, stdout, stderr) => {
+                if (err) {
+                    console.error(`${filePath}: ${err}`);
+                }
+                else {
+                    // Print the *entire* stdout and stderr (buffered).
+                    stdout && console.log(`${filePath}: ${stdout}`);
+                    stderr && console.log(`${filePath}: ${stderr}`);
+                }
+            });
```
There are some warnings here: 
![image](https://user-images.githubusercontent.com/1976647/175827441-5b367e9b-01dd-4f4b-b0b9-4b19d638aef4.png)

0liver replied:
I've ignored the `no-console` rule for the whole file, but I'm not getting the first warning, so I'm not fixing that.

Piedone replied:
I still get "(ESLint) import/no-unresolved	Unable to resolve path to module 'klaw'." here.

0liver replied:
Well, for the life of me, I don't get the error. I've deep-cleaned the repo, and neither before nor after building NE or the solution do I see the error message you're mentioning.

Can you successfully run the scripts, though?

0liver replied:
If yes, then let's call it a day, shall we? We didn't lint our Gulp etc. files in GE at all, so this is still an improvement.

0liver replied:
OK, I've added a rule suppression for that line now.

Piedone commented on Lombiq.NodeJs.Extensions/Lombiq.NodeJs.Extensions.targets:
```diff
@@ -30,36 +34,90 @@
+        Exclude="$([System.IO.Directory]::GetDirectories($(NodeJsExtensionsNpmPackageSourcePath), 'node_modules'))"
+        Condition="Exists($(NodeJsExtensionsNpmPackageSourcePath))" />
+    </ItemGroup>
+    <RemoveDir Directories="@(NodeJsExtensionsNpmPackageFoldersToDelete)" ContinueOnError="true" />
```
Re `ContinueOnError` see this convo, might be applicable: https://github.com/Lombiq/NPM-Targets/pull/20#discussion_r915075950

0liver replied:
Good point!

Commit:
Improve documentation

Commit:
Better ignore possible errors

Commit:
Disable ESLint rule for console statements

Commit:
Improve documentation

Piedone reviewed (changes requested):
See https://github.com/Lombiq/NodeJs-Extensions/pull/3#discussion_r920308501

Commit:
Fix last ESLint warning
